### DISCUSSION TITLE
i18n for not-found page

### DISCUSSION THREAD

**Discussion by simpel** [vercel/next.js#50518] [App Router] [answered]
Hi all!

I'm using the new app router and the described way of doing i18n, it works fine, I get the locale passed as a param to all pages, except the root not-found page. How can I localize that page? And also, is there a way of having differend not-found pages? I have one at app/ that gets shown all the time, when I add a not-found.tsx at /[locale]/ its not used?

The selected not-found.tsx never gets used.

<img width="336" alt="Screenshot 2023-05-30 at 09 13 02" src="https://github.com/vercel/next.js/assets/232024/423abb6c-2789-4492-bde7-d34f481da947">

**Comment by devjiwonchoi** [comment]
It is natural that your root `not-found.tsx` is not affected by localization since it is out of the `/locale` folder.
Since you already have another `not-found.tsx` inside your `/locale` folder, that can be your "Localized" Not Found page.

Note that you need to call `notFound()` function if you want to use the `not-found.tsx` on the same directory.
> The notFound function allows you to render the [not-found file](https://nextjs.org/docs/app/api-reference/file-conventions/not-found) within a route segment...

Example code:
```
import { notFound } from 'next/navigation';
 
async function fetchUsers(id) {
  const res = await fetch('https://...');
  if (!res.ok) return undefined;
  return res.json();
}
 
export default async function Profile({ params }) {
  const user = await fetchUser(params.id);
 
  if (!user) {
    notFound();
  }
 
  // ...
}
```

Then your "Localized Not Found" page could be used.

```
// [locale/not-found.tsx]

import Link from 'next/link';
 
export default function NotFound() {
  return (
    <div>
      <h2>Not Found</h2>
      <p>Could not find requested resource</p>
      <p>
        View <Link href="/blog">all posts</Link>
      </p>
    </div>
  );
}
```

**Reply by simpel** [reply]
Thanks for answering! But I would expect the not-found.tsx in [locale] to be rendered when I visit /en_GB/some/dir/that/does/not/exist but it does not, it only renders the not-found.tsx at the root level. Any ideas why?

**Reply by devjiwonchoi** [reply, marked as answer]
When you set your `[locale]` folder, it means that the route catches only `/something`.
If you want to set the route as `/en_GB/some/dir/that/does/not/exist`, 
you need to change your `locale` folder name as [Catch-all Segments](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#catch-all-segments), which is `[...locale]`.

Since the `notFound()` function renders within a route segment, your `/en_GB/some/dir/that/does/not/exist` is technically out of range of your `[locale]` folder, which is why the root `not-found.tsx` is being rendered.

In conclusion, change your `[locale]` folder name to `[...locale]`.

### UPDATE

@simpel @nikkizol Apologies for providing an inefficient answer.

I have to update the answer since setting the locale to catch all segments is not the best practice & experience for developers.

![Screenshot 2023-09-01 at 3 03 58 AM](https://github.com/vercel/next.js/assets/120007119/ac823de1-ac4e-4680-a4ba-fa0837f46d98)

#### `/[locale]/login` - `notFound`

![Screenshot 2023-09-01 at 3 05 32 AM](https://github.com/vercel/next.js/assets/120007119/6318393c-3a0d-4acc-9246-bafa5bcfbdb4)

#### `/[locale]/catch-all/segments` - `notFound`

![Screenshot 2023-09-01 at 3 06 00 AM](https://github.com/vercel/next.js/assets/120007119/8037f6bc-02ee-44de-89e1-16943c565acc)

#### `/` - `notFound` (outside locale)

![Screenshot 2023-09-01 at 3 05 41 AM](https://github.com/vercel/next.js/assets/120007119/ce66acd1-7aa6-49f9-b7ff-3283e0e45d43)

**Reply by simpel** [reply]
jesus... can't believe i didnt see that.... thanks!

**Reply by nikkizol** [reply]
@simpel have the same structure but I am getting - `Error: Catch-all must be the last part of the URL.`

**Reply by devjiwonchoi** [reply]
@nikkizol I'm assuming that you've tried to set more route segments after `[...slug]`?

`/[...slug]/another-route/page.tsx` is invalid.

**Reply by nikkizol** [reply]
@devjiwonchoi here is my structure 

![Screenshot 2023-08-31 at 19 48 19](https://github.com/vercel/next.js/assets/45672257/5bb4bd08-3344-45cd-9144-4facf9ea8beb)

**Reply by devjiwonchoi** [reply]
It is because your static route `/login` and `/register` are inside the catch-all segment `[...locale]`.

I'll update the answer.

**Reply by nikkizol** [reply]
oh ok, I just saw that @simpel has `[locale]/quiz` and I assumed that he just changed `[locale]` to `[...locale]`

**Reply by foadtkf** [reply]
@devjiwonchoi bro.. can you share the ss of the final folder structure? I'm still a bit confused about the structure.
Here is my structure which is useless as I cannot translate the notfound page:
![image](https://github.com/vercel/next.js/assets/67990762/907f218f-798c-462a-a83a-7ca10a625510)

**Reply by Baha-mumzwolrd** [reply]
What if the locale is needed inside the root layout?

**Reply by jcyuan** [reply]
Hi, @devjiwonchoi is your solution for next14?

**Reply by eythaann** [reply]
This is still working on next 15: 
![image](https://github.com/user-attachments/assets/d12ffda5-05c0-4510-bf9b-7621e90b6a65)
![image](https://github.com/user-attachments/assets/5c84bd5c-3bd2-44bd-8522-60ed7d614d0a)

**Reply by laksamanaap** [reply]
thankyou @eythaann, its work on me

**Reply by gabrielduete** [reply]
Ty, @eythaann! It worked for me:

<img width="398" height="302" alt="image" src="https://github.com/user-attachments/assets/1fb6e22b-d259-443a-9611-e23e6bbad485" />

**Comment by MasteringBOX** [comment]
Hi! 

This solution works to render the not-found located in the catch-all segment. However, how do we get the params in the NotFound function? In the docs it says it doesn't accept any props so I can't get the locale in the NotFound function, how I'm I supposed to localize the text?

Any help would be appreciated.

**Reply by oemer-aran** [reply]
Did you find a solution?

**Reply by DemaPy** [reply]
i use `next-intl` library and localize text via url automatically like: `/en/files` or `/de/files` and use hook `const t = useTranslations("HomePage");`

**Comment by devjiwonchoi** [comment]
Hey everyone I'm trying to open a feature request for this issue and work on it, and just to confirm, how are you guys currently translating the not-found page?

**Reply by foadtkf** [reply]
I created a page.jsx inside a folder [...rest] which is located in the [lang] folder. The translation is working though but sometimes it shows an error that error.js should be  a client component. Do you have any better solution bro?

**Reply by MasteringBOX** [reply]
I do the same thing. The problem is that this doesn't throw a 404 to the browser, it sends a 200 code. 

@foadtkf if you get that error is because you have an error.js file, that file has to be a client component and you have to add the "use client" directive.

**Reply by foadtkf** [reply]
No, it's not bcz of that. As i'm using server side translation, and nextjs somehow automatically declaring this page.tsx as error.tsx. do you have any better approach for that? It's working fine though but showing error in the terminal

**Reply by PxlSyl** [reply]
Hi! Tried a lot of things, but still unable to translate the not-found page

**Reply by blame2020** [reply]
@MasteringBOX 
> The problem is that this doesn't throw a 404 to the browser, it sends a 200 code.

I have same problem..
Have you already solved this problem?

**Comment by arthurpolon** [comment]
What worked for me was to create a 'Catch all route' inside '[locale]' and simply return 'notFound'

![image](https://github.com/vercel/next.js/assets/48016312/507e90a3-649f-431b-b5cd-7752592d01d5)
![image](https://github.com/vercel/next.js/assets/48016312/3a69c8d0-3e59-4151-9669-23df3dfeb4c0)

This way if the route does not match any other route inside '[locale]' It will match this 'Catch all route' and return 'notFound'

**Reply by MasteringBOX** [reply]
Yeah but that 404 page will not be localized

**Reply by vrushank** [reply]
@MasteringBOX You are right. Facing the same issue where i tried to print params or pathname using following code:

```
const pathName = usePathname()

console.log(pathName)
```  
it prints following:

```
/fr/asdf
/_next/static/css/app/[lang]/styles.css.map
```
Same thing happens with page params as well. I get `_next` in the lang param.

I have also defined following in middleware
```
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    '/((?!api|_next/static|_next/image|favicon.ico).*)'
  ]
}
```

It looks like due to catching all the routes it catches the css route. Do you face the same issue? Have you found any solution to this?

**Reply by GuillaumeCode17** [reply]
Like Magic... thank you a lot. 

here's my structure for further reference.... 
![Like_Magic](https://github.com/vercel/next.js/assets/103037993/8bd890a8-12b7-49f0-96df-b88537a19675)

**Reply by Minf97** [reply]
it does work, but remain some problem ...

I set `proxy` in my `middleware.js` so that when it matches URL and proxy to the destination. But as the result of `[...notFound]` , my `middleware` doesn't work

Have any solution to this problem? Thanks a lot.

**Comment by lohnsonok** [comment]
I solved this by making a redirect in  the app/not-found.tsx (`output: 'export'` support)

![image](https://github.com/vercel/next.js/assets/60504466/839ad8cb-dbaf-4edc-91bf-d4dec4f304da)


And have a not-found page.tsx localized:

app/not-found/page.tsx

![image](https://github.com/vercel/next.js/assets/60504466/4b9c2ef2-2a24-4cc1-99a9-669ae2932a5d)

Be sure to process paths that don't return a valid locale at level 1 to redirect to your default locale.

**Comment by PPHelios** [comment]
I found a solution that worked for me on https://next-intl-docs.vercel.app/docs/environments/error-files , I'm not using the same i18n library but the solution provided is global.

**Reply by foadtkf** [reply]
But it wont work if the translation uses server-side-rendering.

**Comment by eivindml** [comment]
Any progress on this? Seems like a reasonable thing to expect. To get the locale as a prop to NotFound route.

**Comment by iamrenejr** [comment]
Tossing my hat into the ring. Regardless of the means to redirect to it, if what we return is the `not-found.tsx` page, it will not have access to the locale, meaning the error page cannot be localised server side. This forces me to render the error page client side since I need to be able to localise my error page.

**Comment by jbukuts** [comment]
Came up with a somewhat elegant fix for my use case. For some context, I'm exporting the site as static assets currently.

What I did was create a new page next to the catch-all segment called `not-found` within the `[locale]` segment. So it's understood as `/[locale]/not-found`. 

This page is created as a static asset and can be localized for every locale at build as a server component. 

I then created a top-level `not-found.tsx`. However instead of this page rendering anything it will look at the current pathname and extract the locale from it (I have a list of valid locales so non-existent ones fallback to `en`) and redirect. So something like `/en/im-not-real` would redirect to `/en/not-found` and so on. 

Downside to this approach is of course you will no longer see the original URL since it redirects. 

Upside is you get server components for your not found page and inheritance from the layout.

Structure for reference:
<img width="258" alt="Screenshot 2024-04-07 at 4 17 01 PM" src="https://github.com/vercel/next.js/assets/14897074/9003a5f8-1996-4472-b0ce-cc761dea8f21">

Top-level `not-found.tsx`:
```tsx
export default function NotFound() {
  const pathName = usePathname();
  const locale = pathName.split('/')[1];
  const finalLocale = locales.includes(locale) ? locale : DEF_LOCALE;
  return redirect(`/${finalLocale}/not-found`, RedirectType.replace);
}
```

Also, for this site **every page** will have a locale. That's why I make the assumption I can extract it from the pathname.

**Reply by foadtkf** [reply]
what if the default locale is excluded from the URL...?

**Reply by krckyboy** [reply]
What do your layout.tsx files look like? I get hydration errors for `lang`.

**Reply by jbukuts** [reply]
@foadtkf If no locale can be found in the URL than a fallback is used. In my case `en`.

**Reply by jbukuts** [reply]
@krckyboy Top level does nothing but return `children`. Just for metadata. The one within `[locale]` looks like:

```tsx
export default function LocaleLayout({
  children,
  params: { locale }
}: Readonly<LocaleLayoutProps>) {
  unstable_setRequestLocale(locale);

  const messages = useMessages();

  return (
    <html lang={locale}>
      <body>
        <NextIntlClientProvider
          messages={pick(messages, ['UIShell', 'components'])}>
          <UIShell>
            {children}
          </UIShell>
        </NextIntlClientProvider>
      </body>
    </html>
  );
}

```

This is for static asset generation so `unstable_setRequestLocale` is used. 
https://next-intl-docs.vercel.app/docs/getting-started/app-router#add-unstable_setrequestlocale-to-all-layouts-and-pages

**Comment by lorenzo-dallamuta** [comment]
I read through the whole thread without seeing mention of a possible hack which pieces together the other solutions while having a localized custom error page and returning a 404 status code:

the idea is still to have a catch-all route inside the `[locale]` dynamic route segment which should only return the `notFound()` function from `next/navigation`, at that point the `not-found.tsx` page from the root folder is going to be rendered, but since it's being rendered through the components inside the `[locale]` path segment the original url will be maintained

from there you can detect the locale however you prefer (I'm doing it with `next-intl`), but the point is that the url will be something like `https://my-site/**en**/unknown-page`, you'll be responding to the browser with a 404 and the jsx will be whatever you specified in your root level `not-found.tsx` which is gonna have access to the current locale

obviously this looks and feels like a hack, a proper solution with the possibility to set a custom error code would be greatly welcome

**Reply by lorenzo-dallamuta** [reply]
mind that this works for me because I'm redirecting all routes to a localized path with next-intl at the middleware layer (so that the not found page will always be rendered through the [locale] path segment)

**Comment by salman486** [comment]
Below is the solution that is neat and clean and worked for me.

Make sure to put layout and not-found.tsx outside of [locale] directory
<img width="422" alt="image" src="https://github.com/user-attachments/assets/38b9dbc5-5536-4036-9a32-f36bc2a2666e">

middleware.ts // make sure matcher has `/((?!_next|_vercel|.*\\..*).*)` to redirect URLs with that does not have locales to with locales
```
import createMiddleware from "next-intl/middleware";

export default createMiddleware({
  locales: ["en", "fr"],

  defaultLocale: "fr",
});

export const config = {
  matcher: ["/", "/(fr|en)/:path*", "/((?!_next|_vercel|.*\\..*).*)"],
};
```

i18n.ts
```
import { notFound } from "next/navigation";
import { getRequestConfig } from "next-intl/server";
import { locales } from "./config";

export default getRequestConfig(async ({ locale }) => {
  if (!locales.includes(locale as any)) notFound();

  return {
    messages: (await import(`./messages/${locale}.json`)).default,
  };
});
```

next.config.mjs
```
import createNextIntlPlugin from "next-intl/plugin";

const withNextIntl = createNextIntlPlugin("./src/i18n.ts");

/** @type {import('next').NextConfig} */
const nextConfig = {};

export default withNextIntl(nextConfig);
```

**Comment by aapostolou** [comment]
For anyone looking for a solution for this on `NextJS 15+`, i have found a solution

First off, on the middleware i added the current path to the headers in order to use it later to get the locale

```tsx
// middleware.ts
...
const headers = new Headers(request.headers)
headers.set('x-current-path', request.nextUrl.pathname)
...
```

After setting the headers i created a function called `getLocaleFromPathname`, which checks the path for the locale or uses the default one if it doesn't exists. (similar to what you are probably doing on your middleware)

```tsx
// utils/getLocaleFromPathname.tsx
import { headers } from 'next/headers'

import { DictionaryLanguages } from '@/dictionaries/dictionaries' // This is just the type of the locales - 'en' | 'gr'

export const locales: string[] = ['en', 'gr'] as const
export const defaultLocale: (typeof locales)[number] = 'en'

export const getLocaleFromPathname = async (): Promise<DictionaryLanguages> => {
  const headerList = await headers()
  const pathname = headerList.get('x-current-path') ?? ''

  const locale =
    locales.find(
      (locale) =>
        pathname.startsWith(`/${locale}/`) || pathname === `/${locale}`
    ) ?? defaultLocale

  return locale as DictionaryLanguages
}
```

I move both `layout.tsx` and `not-found.tsx` to the root (under `/app`)

Now on both of them you can just run the `getLocaleFromPathname` function instead of getting the locale from the `params` prop

```tsx
// layout.tsx
import { getLocaleFromPathname } from '@/utils/getLocaleFromPathname'

export default async function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode
}>) {
  const locale = await getLocaleFromPathname()

  return (
    <html lang={locale}>
      <body>
        {children}
      </body>
    </html>
  )
}
```

```tsx
// not-found.tsx
import { getDictionary } from '@/dictionaries'
import { getLocaleFromPathname } from '@/utils/getLocaleFromPathname'

export default async function NotFound() {
  const locale = await getLocaleFromPathname()
  const dict = await getDictionary(locale)

  return (
    <div>
      <h1>{dict['not-found'].title}</h1>
    </div>
  )
}

```

Hope this helps. Happy Coding !

**Reply by vanska** [reply]
This solution means every route will be [opted into dynamic rendering](https://nextjs.org/docs/app/api-reference/functions/headers#good-to-know) to read the header thus braking any static behaviour.

**Reply by funkeeflow** [reply]
Why not just use the approach for the not-found.tsx. layout.tsx can receive params, thus putting it inside [locale] would work.

**Reply by nagman** [reply]
Just tried it and it works like a charm :)
This is the best solution so far.
Here's the doc on how to properly add a request header: https://nextjs.org/docs/app/building-your-application/routing/middleware#setting-headers (you need to return the new request headers in the middleware)
I've used it only on the 404 page, leaving the layout as is.

**Comment by allanice001** [comment]
simplest solution i found:
remove root level not-found page, only keep the localised not found
```
src/app
├── [locale]
│   ├── [...rest]
│   │   └── page.tsx
│   ├── layout.tsx
│   └── not-found.tsx
├── favicon.ico
└── globals.css
```

the catchall page is as simple as 
```
import { notFound } from "next/navigation"

export default function CatchAllPage() {
  notFound()
}
```
as long as your middleware handles localization propperly (whether its in the path `example.com./en/fakepage` as well as `example.com/fakepage`, or a subdomain `en.example.com` you should see your localised error page

this example is straight from https://next-intl.dev/docs/environments/error-files#not-foundjs

**Comment by gmonte** [comment]
This router tree worked for me:

![image](https://github.com/user-attachments/assets/6dbebcf8-0028-473d-8e78-a35c6f5dba90)

```typescript
// intl-config.ts

export const locales = ["en", "pt-BR"] as const;

export type Locale = (typeof locales)[number];

export const intlConfig = {
  locales,
  defaultLocale: "en" as Locale
};
```

```typescript
// middleware.ts

import type { NextRequest } from "next/server";
import { i18nRouter } from "next-i18n-router";

export default async function middleware(request: NextRequest) {
    return i18nRouter(request, intlConfig);
    
    // if you need to validate `api` with another middleware
    // comment the return above, include "/api(.*)" into the matcher,
    // and skip the i18n router for it.
     
    if (!request.nextUrl.pathname.startsWith("/api")) {
      // we don't want to use i18nRouter for api routes
      return i18nRouter(request, intlConfig);
    }
}

export const config = {
  matcher: [
    // Skip Next.js internals and all static files, unless found in search params
    "/((?!_next|_vercel|.*\\..*).*)",
    // "/api(.*)" // 👈 uncomment it if you need to validate api with another middleware
  ],
};

```

```typescript
// app/not-found.tsx

"use client";

import Error from "next/error";

export default function NotFound() {
  return (
    <html lang="en">
      <body>
        <Error statusCode={404} />
      </body>
    </html>
  );
}
```

```typescript
// app/layout.tsx

// Since we have a root `not-found.tsx` page, a layout file
// is required, even if it's just passing children through.
export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return children;
}
```

```typescript
// app/[locale]/layout.tsx

import { locales, type Locale } from "@/intl-config.ts`

export default async function AppLayout({
  children,
  params,
}: Readonly<{
  children: React.ReactNode;
  params: Promise<{ locale: Locale }>;
}>) {
  const { locale } = await params;

  if (!locales.includes(locale)) {
    notFound();
  }

  // here you are safe to use intl 
  
  return (
    <html lang={locale}>
      ...
    </html>
  )
}
```

```typescript
// app/[locale]/[...rest]/page.tsx

import { notFound } from "next/navigation";

export default function CatchAllPage() {
  notFound();
}
```

```typescript
// app/[locale]/not-found.tsx

import { redirect } from "next/navigation";

export default function NotFoundPage() {
  return (
    <div>
      do your i18n magic here
    </div>
  )
}
```
